### PR TITLE
Support for Odroid C2 and Arch Linux.

### PR DESCRIPTION
**Pull request by larmog**
Uses `aufs` and works with `armhf` images, aarch64 gives `aufs` errors . Tried `devicemapper` but got strange errors, XFS support in kernel is experimental, maybe works with `loop-lvm`.

### CODE REVIEW

File: sdcard/os/archlinux.sh [hunk 1]
```diff
@@ -21,6 +23,8 @@ initos(){
             generaldownload;;
         cubietruck|bananapro)
             allwinnerdownload;;
+        odroid-c2)
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: I think you can remove `amlogicdownload` and use `generaldownload` instead

File: sdcard/os/archlinux.sh [hunk 2]
```diff
@@ -154,4 +160,41 @@ allwinnerdownload(){
...
+n
+p
+1
+2048
+$SDCARDSIZE
+w
+EOF
+
+	# 4. Create and mount the ext4 filesystem:
+	mkfs.ext4 ${SDCARD}1
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: Use `$PARTITION1` instead

File: sdcard/os/archlinux.sh [hunk 3]
```diff
@@ -154,4 +160,41 @@ allwinnerdownload(){
...
+p
+n
+p
+1
+2048
+$SDCARDSIZE
+w
+EOF
+
+	# 4. Create and mount the ext4 filesystem:
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: What happened to 3? :smile:

File: sdcard/os/archlinux.sh (hunk 1 above)
Comment (larmog) [inline sdcard/os/archlinux.sh]: Odroid needs `bsdtar`.

File: sdcard/os/archlinux.sh (hunk 2 above)
Comment (larmog) [inline sdcard/os/archlinux.sh]: Ok.

File: sdcard/os/archlinux.sh (hunk 1 above)
Comment (luxas) [inline sdcard/os/archlinux.sh]: Why? What's the difference?

File: sdcard/os/archlinux.sh [hunk 4]
```diff
@@ -154,4 +160,47 @@ allwinnerdownload(){
 
 umount_root(){
     umount $ROOT
-}
\ No newline at end of file
+}
+
+amlogic(){
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: amlogicformat

File: sdcard/os/archlinux.sh [hunk 5]
```diff
@@ -154,4 +160,47 @@ allwinnerdownload(){
...
+	# 4. Create and mount the ext4 filesystem:
+	mkfs.ext4 $PARTITION1
+	mount $PARTITION1 $ROOT
+}
+
+amlogicdownload(){
+	# Install bsdtar if not present
+	require bsdtar
+
+	ARCH_BOARD=$MACHINENAME
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: Why not use `$MACHINENAME` directly?

File: sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-flannel.conf [hunk 6]
```diff
@@ -15,7 +15,8 @@ ExecStart=/usr/bin/docker daemon \
     --insecure-registry=registry.kube-system.svc.cluster.local:5000 \
     --insecure-registry=10.0.0.20:5000 \
     --insecure-registry=registry.kube-system:5000 \
-    --exec-opt native.cgroupdriver=cgroupfs
+    --exec-opt native.cgroupdriver=cgroupfs \
+    $DOCKER_STORAGE_OPTS
```
Comment (luxas) [inline sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-flannel.conf]: `${DOCKER_STORAGE_OPTS}`

File: sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-overlay.conf [hunk 7]
```diff
@@ -7,7 +7,8 @@ EnvironmentFile=/etc/kubernetes/k8s.conf
 ExecStart=
 ExecStart=/usr/bin/docker daemon \
     -H unix:///var/run/docker.sock \
```
Comment (luxas) [inline sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-overlay.conf]: Set to `--host` so all args are `--`

File: sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-overlay.conf [hunk 8]
```diff
@@ -7,7 +7,8 @@ EnvironmentFile=/etc/kubernetes/k8s.conf
...
-    -s ${DOCKER_STORAGE_DRIVER} \
+    --storage-driver ${DOCKER_STORAGE_DRIVER} \
     --exec-opt native.cgroupdriver=cgroupfs
+    $DOCKER_STORAGE_OPTS
```
Comment (luxas) [inline sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-overlay.conf]: dito

File: sdcard/rootfs/kube-systemd/usr/lib/systemd/system/system-docker.service [hunk 9]
```diff
@@ -8,14 +8,15 @@ Type=notify
 EnvironmentFile=/etc/kubernetes/k8s.conf
 ExecStart=/usr/bin/docker daemon \
     --host=unix:///var/run/system-docker.sock \
-    --storage-driver=${DOCKER_STORAGE_DRIVER} \
+    --storage-driver ${DOCKER_STORAGE_DRIVER} \
```
Comment (luxas) [inline sdcard/rootfs/kube-systemd/usr/lib/systemd/system/system-docker.service]: Why this change?

File: sdcard/rootfs/kube-systemd/usr/lib/systemd/system/system-docker.service [hunk 10]
```diff
@@ -8,14 +8,15 @@ Type=notify
...
     --exec-opt native.cgroupdriver=cgroupfs \
     --exec-root=/var/run/system-docker \
     --pidfile=/var/run/system-docker.pid \
     --iptables=false \
     --ip-masq=false \
     --bridge=none \
-    --graph=/var/lib/system-docker
+    --graph=/var/lib/system-docker \
+    $DOCKER_STORAGE_OPTS
```
Comment (luxas) [inline sdcard/rootfs/kube-systemd/usr/lib/systemd/system/system-docker.service]: dito

File: sdcard/write.sh [hunk 11]
```diff
@@ -139,11 +140,11 @@ if [[ -z $QUIET || $QUIET == 0 ]]; then
...
     read -p "You are going to lose all your data on $1. Continue? (Y is default) [Y/n]" answer
 
-    case $answer in 
-        [nN]*) 
+    case $answer in
+        [nN]*)
             echo "Quitting..."
             rm -r $TMPDIR
-            exit 1;;        
+            exit 1;;
```
Comment (luxas) [inline sdcard/write.sh]: Appreciated that you clean up unneeded spaces :)

**Comment by luxas** [comment]
Okay, let's continue on this one now when `v0.7.0` is out

File: sdcard/os/archlinux.sh (hunk 1 above)
Comment (larmog) [inline sdcard/os/archlinux.sh]: Don't know for sure [see.](http://unix.stackexchange.com/questions/101561/what-are-the-differences-between-bsdtar-and-gnu-tar). I obviously tried standard `tar`, but it failed. So I switched to the archlinux instructions.

**Comment by dogieBert** [comment]
I am currently trying to port this to the Odroid U3 ( Which will add U2 Support aswell )
As soon as this gets merged, I will post a PR based on it ;)

File: sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-flannel.conf (hunk 6 above)
Comment (larmog) [inline sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-flannel.conf]: This should be `DOCKER_OPTS` rather, as you can add any docker option. The environment env should be passed as is without any escaping ${ENV} is handled differently than $ENV by systemd.

File: sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-overlay.conf (hunk 8 above)
Comment (larmog) [inline sdcard/rootfs/kube-systemd/etc/kubernetes/dropins/docker-overlay.conf]: Should be $DOCKER_OPTS rather and `--storage-opt` and can be configured for different boards. If it's not set then it makes no harm.

File: sdcard/rootfs/kube-systemd/usr/lib/systemd/system/system-docker.service (hunk 9 above)
Comment (larmog) [inline sdcard/rootfs/kube-systemd/usr/lib/systemd/system/system-docker.service]: A miss, sorry. Leftover from some experiment.

**Comment by luxas** [comment]
@dogieBert Good to hear that :)

@larmog Let's finish this work now, and merge it.

File: sdcard/os/archlinux.sh [hunk 12]
```diff
@@ -155,3 +161,45 @@ allwinnerdownload(){
 umount_root(){
     umount $ROOT
 }
+
+amlogicformat(){
+
+  # See https://archlinuxarm.org/platforms/armv8/amlogic/odroid-c2
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: indentation

File: sdcard/os/archlinux.sh [hunk 13]
```diff
@@ -155,3 +161,45 @@ allwinnerdownload(){
...
+	# 5. Download and extract the root filesystem
+	# Download, redirect stderr (all errors) to stdout, which in turn is appended to a log file
+	curl -sSL -k http://archlinuxarm.org/os/ArchLinuxARM-${MACHINENAME}-latest.tar.gz | bsdtar -xpf - -C $ROOT >> $LOGFILE 2>&1
+}
+
+write_boot_loader_and_umount_root(){
+	# 6. Flash the bootloader files
+	CURRDIR=`pwd`
+  cd $ROOT/boot
+  ./sd_fusing.sh $SDCARD
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: document something about what this is doing

File: sdcard/rootfs/kube-systemd/etc/kubernetes/dynamic-env/board/odroid-c2.sh [hunk 14]
```diff
@@ -0,0 +1,6 @@
+# This command is run by kube-config when doing "kube-config install"
+board_post_install(){
+  echo "Board post install ..."
+  sed -e "s@overlay@devicemapper@" -i $KUBERNETES_CONFIG
+  echo "DOCKER_OPTS=\"--storage-opt dm.fs=ext4\"" >> $KUBERNETES_CONFIG
```
Comment (luxas) [inline sdcard/rootfs/kube-systemd/etc/kubernetes/dynamic-env/board/odroid-c2.sh]: please `echo 'DOCKER_OPTS="--storage-opt dm.fs=ext4"'` for readability

File: sdcard/os/archlinux.sh (hunk 1 above)
Comment (luxas) [inline sdcard/os/archlinux.sh]: tried `curl -sSL -k http://archlinuxarm.org/os/ArchLinuxARM-odroid-c2-latest.tar.gz | tar -xz -C odroid-c2` without problems on an archlinux host. Works without problems

Of course there are the `tar: Ignoring unknown extended header keyword 'SCHILY.fflags'` warnings, but one may ignore them

File: sdcard/os/archlinux.sh [hunk 15]
```diff
@@ -155,3 +161,45 @@ allwinnerdownload(){
...
+  cd $CURRDIR
+	# 8. Unmount the partition (#7 Is optional and not needed)
+  umount $ROOT
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: You may call `umount_root` here also

File: sdcard/os/archlinux.sh [hunk 16]
```diff
@@ -155,3 +161,45 @@ allwinnerdownload(){
...
     umount $ROOT
 }
+
+amlogicformat(){
+
+  # See https://archlinuxarm.org/platforms/armv8/amlogic/odroid-c2
+	# 1. Zero the beginning of the SD card or eMMC module:
+	dd if=/dev/zero of=$SDCARD bs=1M count=8
+
+	# 2. and 3. Start fdisk to partition the SD card:
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: Turns out that this code (2,3,4) is equal to `allwinnerformat`.
`allwinnerformat` should be renamed to `onepartitionformat` and `generalformat` to `twopartitionformat`

**Comment by larmog** [comment]
I don't have all the boards so that I can verify that it works. I've ordered new `odroid-c2` boards and I can verify that `odroid-c2`, `rpi` and `rpi-2` works.

I'll squash all commits after all has been tested. Ok?

File: sdcard/os/archlinux.sh [hunk 17]
```diff
@@ -52,7 +54,10 @@ checkrootfs(){
 
 # Format the SD Card with two partitions, boot and root. Boot is $1 MB big.
 # This makes the root partition as big as the sd card minus boot
-generalformat(){
+twopartitionformat(){
+    if [[ ! -z $SDCARDSIZE ]]; then
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: actually, I removed these as they're not longer required, so you may safely remove them

File: sdcard/os/archlinux.sh [hunk 18]
```diff
@@ -158,3 +166,14 @@ allwinnerdownload(){
...
+
+write_boot_loader_and_umount_root(){
+    CURRDIR=`pwd`
+    cd $ROOT/boot
+
+    # Flash the bootloader files
+    ./sd_fusing.sh $SDCARD
+    cd $CURRDIR
+
+    umount_root()
```
Comment (luxas) [inline sdcard/os/archlinux.sh]: no `()` at the end I think

**Comment by luxas** [comment]
Thanks @larmog for the work :smile:! 
Only two nits before merge

**Comment by larmog** [comment]
@luxas Sorry for my unclear comment earlier today (removed). Just to be clear, I haven't had time to test this. My `odroid` board is occupied and I'm hoping that the new boards will arrive this week, so that I can test the change. If you feel confident, merge the PR and I'll verify that it work as soon as possible.

**Comment by luxas** [comment]
LGTM, thanks for the contribution!

**Comment by larmog** [comment]
The PR is broken. `archlinux.sh:111` does not work with `odroid-c2`.
